### PR TITLE
Add NativeTest for BigInt class

### PR DESCRIPTION
**Pull request by duongnhn**
- Create first nativetests for BigInt class in DataStructure.
- Cover: Init, Add, Substract.
- TODO: add more for other operation.

### CODE REVIEW

File: bin/NativeTests/BigIntTest.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+template <typename EncodedChar>
+double Js::NumberUtilities::StrToDbl(const EncodedChar *, const EncodedChar **, bool& )
+{
+    assert(false);
+    return 0.0;// don't care
+}
+
+
+#include "..\..\lib\Common\DataStructures\BigInt.h"
+#include "..\..\lib\Common\DataStructures\BigInt.cpp"
```
Comment (jackhorton) [inline bin/NativeTests/BigIntTest.h]: I believe the preferred method for exposing internals like this to tests is using TestHooks -- it may make some of the code here cleaner, though it may add its own complexity of requiring an ad-hoc C interface to the data structure. @curtisman thoughts?

File: lib/Common/DataStructures/BigInt.h [hunk 2]
```diff
@@ -50,8 +51,8 @@ namespace Js
         void Subtract(BigInt *pbi);
         int DivRem(BigInt *pbi);
 
-        int32 Clu(void);
-        uint32 Lu(int32 ilu);
```
Comment (jackhorton) [inline lib/Common/DataStructures/BigInt.h]: How much existing code uses this data structure? Is it possible that we could rename these methods to be more reasonable without changing a ton of unrelated code to match?

File: bin/NativeTests/BigIntTest.cpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        f = bi1.FInitFromRglu(digits, length);
+        Assert(f);
+        f = bi2.FInitFromRglu(digits, length);
+        Assert(f);
+        // Equal
+        int result;
+        result = bi1.Compare(&bi2);
+        CHECK(result == 0);
+        f = bi2.FInitFromBigint(&bi1);
+        Assert(f);
```
Comment (jackhorton) [inline bin/NativeTests/BigIntTest.cpp]: A slightly more Catch-y way to do this is to use REQUIRE() (REQUIRE stops the program when the condition fails, CHECK allows the execution to continue)

File: lib/Common/DataStructures/BigInt.h (hunk 2 above)
Comment (duongnhn) [reply lib/Common/DataStructures/BigInt.h]: Pretty much the code here is used to convert a string to double. I found about 30-40 references to this code.

---
In reply to: [214254468](https://github.com/Microsoft/ChakraCore/pull/5657#discussion_r214254468) [](ancestors = 214254468)

File: bin/NativeTests/BigIntTest.h (hunk 1 above)
Comment (duongnhn) [reply bin/NativeTests/BigIntTest.h]: Yes, TestHooks is a good approach too! I might refactor this class to make it more readable (in other PR) or create a new class. Haven't decided yet

---
In reply to: [214253986](https://github.com/Microsoft/ChakraCore/pull/5657#discussion_r214253986) [](ancestors = 214253986)

**Comment by duongnhn** [comment]
@botnet test Windows 8.1 arm_release please

**Comment by MSLaguana** [comment]
@dotnet-bot test Windows 8.1 arm_release please

**Comment by MSLaguana** [comment]
Looks like a legit prefast failure:
```
15:45:14 d:\j\workspace\arm_release_p---1a297bd3\bin\nativetests\biginttest.h(20): warning C28251: Inconsistent annotation for 'ReportAssert': this instance has no annotations. See d:\j\workspace\arm_release_p---1a297bd3\lib\common\exceptions\throw.h(38).  [D:\j\workspace\arm_release_p---1a297bd3\bin\NativeTests\NativeTests.vcxproj]
```

File: bin/NativeTests/BigIntTest.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+#include "..\..\lib\Common\Common\NumberUtilities.cpp"
+
+namespace Js
+{
+    void Throw::FatalInternalError(long)
+    {
+        Assert("Not tested");
+    }
+
+    bool Throw::ReportAssert(char const *, unsigned int, char const *, char const *)
```
Comment (MSLaguana) [inline bin/NativeTests/BigIntTest.h]: This exists elsewhere as 
`bool Throw::ReportAssert(__in LPCSTR fileName, uint lineNumber, __in LPCSTR error, __in LPCSTR message)`

Why are you duplicating it? And why do you `Assert("Not tested")` which will always pass?

Comment (duongnhn) [reply bin/NativeTests/BigIntTest.h]: I am mocking it. But you are right. I should assert(false) instead.

---
In reply to: [214491945](https://github.com/Microsoft/ChakraCore/pull/5657#discussion_r214491945) [](ancestors = 214491945)

Comment (MSLaguana) [reply bin/NativeTests/BigIntTest.h]: If you are stubbing it, then you'll need to match the prefast annotations with `__in` then or the build will complain as it is now. Why are you stubbing this out though? If these are invoked, then won't it fail just as hard as `Assert(false)` would?
